fix(printing): handle python printing of piecewise with "assign_to"

ccam80 opened the pull request:
#### References to other Issues or PRs
None

#### Brief description of what is fixed or changed
When printing an assignment where the RHS is a piecewise assignment, the current PythonCodePrinter prints two assignments in the ternary if-else form, which is invalid Python. This commit adds a check in _print_Assignment which tests whether the RHS is a Piecewise expression, and if so, returns "lhs = rhs", letting the Piecewise operator be correctly processed as an expression in the next round.

Example:
```python
from sympy import symbols, Piecewise
from sympy.printing.pycode import PythonCodePrinter

x, k, c = symbols('x k c')
p = PythonCodePrinter()
piecewise_expr = Piecewise((x, k), (0, True))
pycode_assign_to = p.doprint(piecewise_expr, assign_to=x)

print("Assign to:        ", pycode_assign_to)
print("Expected:   ", "x = (k) if k else (0))")
```
#### Other comments
I don't *think* this shortcuts any important steps elsewhere in the printers, however I'm not super familiar with Sympy. All python and pycode printing tests are passing with this edit on my machine. Please let me know if a different approach is required!

#### Release Notes
<!-- BEGIN RELEASE NOTES -->
* printing
  * Fixed handling of assignments from Piecewise expressions in PythonCodePrinter
<!-- END RELEASE NOTES -->

ccam80 committed:
fix(printing): handle python printing of piecewise with "assign_to"

When printing an assignment where the RHS is a piecewise assignment, the current PythonCodePrinter prints two assignments in the ternary if-else form, which is invalid Python. This commit adds a check in _print_Assignment which tests whether the RHS is a Piecewise expression, and if so, returns "lhs = rhs", letting the Piecewise operator be correctly processed as an expression in the next round.

Example:
>>> from sympy import symbols, Piecewise
>>> from sympy.printing.pycode import PythonCodePrinter
>>> x, k, c = symbols('x k c')
>>> p = PythonCodePrinter()
>>> piecewise_expr = Piecewise((x, k), (0, True))
>>> pycode_assign_to = p.doprint(piecewise_expr, assign_to=x)
>>> print("Assign to:        ", pycode_assign_to)
>>> print("Expected:   ", "x = (k) if k else (0))")

ccam80 committed:
chore: add github noreply address to mailmap

bjodah commented:
Nice catch. I'm trying to figure out why it prints like that in the first place, these lines look suspicious to me:
https://github.com/sympy/sympy/blob/7c29689e928313a31af2e525c2f76f33b62576f0/sympy/printing/codeprinter.py#L378-L386

I wonder if that is ever useful? Did you try patching those lines directly?

bjodah reviewed (changes requested):
Apart from my minor requests, I think this looks good!

ccam80 committed:
fix: Move piecewise printing fix to AbstractPythonPrinter

PR #28371 identified and fixed a bug in piecewise printing using the pycode printer, but patched PythonCodePrinter and so left some printers buggy. This commit moves the patch to AbstractPythonPrinter. The regression test has been updated to also test SymPyPrinter, confirming that the whole family of pycode printers are no longer bugged.

ccam80 commented:
Thanks for the speedy review! AbstractPythonCodePrinter makes more sense as the home for this patch, and adding a SymPyPrinter test confirms that the fix flows into the other concrete pycode printers. Let me know if there's any additional edits that might make this PR more useful.

Multiple assignments makes sense in a multiline conditional. It looks as the decision to use ternary form versus multiline is printer- rather than complexity-dependent:

```python
from sympy import symbols, Piecewise
from sympy.printing.pycode import PythonCodePrinter
from sympy.printing.c import C99CodePrinter

x, k, c = symbols('x k c')
p = PythonCodePrinter()
cc99 = C99CodePrinter()
multi_piecewise = Piecewise((x, k), (x**2, c), (x, x>c), (k,x<=k), (0, True))

nested_ternaries_pycode = p.doprint(multi_piecewise, assign_to=x)
multiline_conditionals_ccode = cc99.doprint(multi_piecewise, assign_to=x)

print("Nested ternaries: \n", nested_ternaries_pycode)
print("Multiline:        \n", multiline_conditionals_ccode)
```

In all(?) of the code printer languages supported there's some form of ternary-like syntax for conditionals. It might make more sense from a readability perspective to print one-line ternary-form conditionals for two-option Piecewise functions, and fall back to multiline printing for >=3 pieces. If this sounds like something that would improve Sympy, I'd be happy to continue diving into the code-printer world and put a new PR together when I get a bit more time.

ccam80 commented on sympy/printing/pycode.py:
```diff
@@ -592,6 +592,16 @@ def _print_Symbol(self, expr):
         else:
             return name
 
+    def _print_Assignment(self, expr):
```
Moved to AbstractPythonCodePrinter.

bjodah commented:
Thank you @ccam80

Ah, yes. I think the issue is fortran: it doesn't support a ternary operator I believe.
